File: samples/replicatedentity-counter/src/main/java/com/example/counter/domain/AbstractCounter.java
// FIXME codegen for ReplicatedEntity

package com.example.counter.domain;

import com.akkaserverless.javasdk.replicatedentity.ReplicatedEntity;
import com.akkaserverless.javasdk.replicatedentity.ReplicatedCounter;
import com.example.counter.CounterApi;
import com.google.protobuf.Empty;

public abstract class AbstractCounter extends ReplicatedEntity<ReplicatedCounter> {

  public abstract Effect<Empty> increase(ReplicatedCounter currentData, CounterApi.IncreaseValue command);

  public abstract Effect<Empty> decrease(ReplicatedCounter currentData, CounterApi.DecreaseValue command);

  public abstract Effect<CounterApi.CurrentCounter> getCurrentCounter(ReplicatedCounter currentData, CounterApi.GetCounter command);

}

File: samples/scala-replicated-examples/src/main/scala/com/example/replicated/counter/AbstractCounter.scala
package com.example.replicated.counter

import com.akkaserverless.scalasdk.replicatedentity.ReplicatedCounter
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedDataFactory
import com.akkaserverless.scalasdk.replicatedentity.ReplicatedEntity
import com.google.protobuf.empty.Empty

abstract class AbstractCounter extends ReplicatedEntity[ReplicatedCounter] {

  override def emptyData(factory: ReplicatedDataFactory): ReplicatedCounter =
    factory.newCounter

  def increase(counter: ReplicatedCounter, increaseValue: IncreaseValue): ReplicatedEntity.Effect[Empty] =
    effects
      .update(counter.increment(increaseValue.value))
      .thenReply(Empty.defaultInstance)

  def decrease(counter: ReplicatedCounter, decreaseValue: DecreaseValue): ReplicatedEntity.Effect[Empty] =
    effects
      .update(counter.decrement(decreaseValue.value))
      .thenReply(Empty.defaultInstance)

  def get(counter: ReplicatedCounter, getValue: GetValue): ReplicatedEntity.Effect[CurrentValue] =
    effects.reply(CurrentValue(counter.value))

}
